File: Swift/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.swift
import Foundation

func solution(_ record:[String]) -> [String] {
    let splitedRecord = record.map({ $0.split(separator: " ") })
    var userNick: [String: String] = [:]
    var messages: [(String, Bool)] = []
    var result: [String] = []
    
    for recordLine in splitedRecord {
        let command = String(recordLine[0])
        let userID = String(recordLine[1])
        if command == "Enter" {
            userNick[userID] = String(recordLine[2])
            messages.append((userID, true))
        } else if command == "Leave" {
            messages.append((userID, false))
        } else if command == "Change" {
            userNick[userID] = String(recordLine[2])
        }
    }
    
    for message in messages {
        if let nickName = userNick[message.0] {
            if message.1 {
                result.append("\(nickName)님이 들어왔습니다.")
            } else {
                result.append("\(nickName)님이 나갔습니다.")
            }
        }
    }
    return result
}

File: Java/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.HashMap;
import java.util.ArrayList;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String, String> user = new HashMap<>();
        
        for (String commandLine: record) {
            String[] splited = commandLine.split(" ");
            if (splited[0].equals("Enter") || splited[0].equals("Change")) {
                user.put(splited[1], splited[2]);
            }
        }
        
        ArrayList<String> message = new ArrayList<>();
         for (String commandLine: record) {
            String[] splited = commandLine.split(" ");
            if (splited[0].equals("Enter")) {
                message.add(user.getOrDefault(splited[1], "")+"님이 들어왔습니다.");
            } else if (splited[0].equals("Leave")) {
                message.add(user.getOrDefault(splited[1], "")+"님이 나갔습니다.");
            }
        }
        return message.toArray(new String[0]);
    }
}